[feature] Implemented config app API filters #747

Aryamanz29 opened the pull request:
![Screenshot from 2023-03-14 17-43-46](https://user-images.githubusercontent.com/56113566/225000464-36c9441b-36b6-4622-8c03-6286c9413bee.png)

Closes #747

#### Todo

- [x]  Add docs
- [x]  Add API filters tests.

nemesisdesign commented on openwisp_controller/config/api/filters.py:
```diff
@@ -0,0 +1,100 @@
+        return queryset.exclude(devicelocation__isnull=value)
+
+    def _set_valid_filterform_lables(self):
+        self.filters['config__status'].label = 'Config status'
```
flag as translatable this and any other label

nemesisdesign commented on openwisp_controller/config/api/filters.py:
```diff
@@ -0,0 +1,100 @@
+    def _set_valid_filterform_lables(self):
+        self.filters['config__status'].label = 'Config status'
+        self.filters['config__backend'].label = 'Config backend'
+        self.filters['devicelocation'].label = 'DeviceLocation'
```
```suggestion
        self.filters['devicelocation'].label = 'Device location'
```

nemesisdesign commented on openwisp_controller/config/api/filters.py:
```diff
@@ -0,0 +1,100 @@
+
+class DeviceListFilter(BaseConfigAPIFilter):
+    config__templates = filters.CharFilter(
+        field_name='config__templates', label='Config template'
```
these labels too, all strings which are shown in the UI have to be flagged as translatable

nemesisdesign reviewed (changes requested):
The first impression is good but a basic test for each filter is missing, we don't need anything fancy, the simplest the better, but the test must spot if the filter is removed or broken in a significant way (eg: if you remove a specific filter the subtest/assertion must fail).

Please remember to flag labels as translatable.

nemesisdesign commented on openwisp_controller/config/api/filters.py:
```diff
@@ -0,0 +1,103 @@
+class DeviceListFilter(BaseConfigAPIFilter):
+    config__templates = filters.CharFilter(
+        field_name='config__templates', label=_('Config template')
+    )
```
I tried passing two UUIDs, comma separated, to see if it handled and the API returned an internal server error for invalid UUID. Please add a test for this and make sure the API returns 400 bad request.

Aryamanz29 replied:
sure

nemesisdesign commented on README.rst:
```diff
@@ -1427,6 +1427,66 @@ List devices
+
+.. code-block:: text
+
+   GET /api/v1/controller/device/?devicelocation={devicelocation}
```
this in the admin is called `with_geo`, can we maintain consitency please?

Aryamanz29 replied:
sure

nemesisdesign commented on README.rst:
```diff
@@ -1637,6 +1697,23 @@ List device groups
+
+.. code-block:: text
+
+   GET /api/v1/controller/group/?device={device}
```
This in the admin is called `empty=`, can we maintain consistency please?

Aryamanz29 replied:
sure

nemesisdesign reviewed (changes requested):
Great progress Aryaman!

Please log in to the demo system and look at these endpoints:

- https://demo.openwisp.io/api/v1/controller/floorplan/
- https://demo.openwisp.io/api/v1/controller/location/
- https://demo.openwisp.io/api/v1/network-topology/node/

For all relations, can you add filters which show the available options in the UI? See the network topology example, eg:

For organization, can you try adding both organization as a relation and organization slug (as floorplan and location endpoints)? This should make it easier for the user to use whatever they have (whether it's the org ID or org slug).

We have added multi-tenant capabilities in these filters already if I am not mistaken, please read the and nearby sections extensively: https://github.com/openwisp/openwisp-users#django-rest-framework-mixins .

Aryamanz29 commented:
> For all relations, can you add filters which show the available options in the UI? See the network topology example, eg:

- If we add available options for the relation in the filter form using something other than a `CharFilter` (i.e., an input field), it will result in adding an extra SQL duplicate query to the endpoint. This is because the same set of queries have already been executed by DRF earlier to show the create form and display available options; see the example below.

### LIST DEVICE

#### Currently, there are no relation options in the filter form:

![Screenshot from 2023-03-17 00-30-46](https://user-images.githubusercontent.com/56113566/225726529-ba8e4e7d-f21c-473b-86ff-d4a63ca4d0f4.png)

#### After adding options to the filter form (added options to `template` and `device` groups):

![Screenshot from 2023-03-17 00-21-27](https://user-images.githubusercontent.com/56113566/225726035-a91ecf30-8f74-4110-ae23-e4cdeae59cc9.png)

> For organization, can you try adding both organization as a relation and organization slug (as floorplan and location endpoints)? This should make it easier for the user to use whatever they have (whether it's the org ID or org slug).
> 
> We have added multi-tenant capabilities in these filters already if I am not mistaken, please read the and nearby sections extensively: [openwisp/openwisp-users#django-rest-framework-mixins](https://github.com/openwisp/openwisp-users#django-rest-framework-mixins) .

Sure, we can add both `organization_id` and `organization_slug` filters. For that, we can add the organization field to `BaseOrganizationSlugFilter` of `openwisp_controller.geo.api.views` and reuse that logic in `BaseConfigAPIFilter` (added in [`9b5e243` (#748)](https://github.com/openwisp/openwisp-controller/pull/748/commits/9b5e243b27d66137a2fbeb58067a6fe71cc0df55))

nemesisdesign commented on openwisp_controller/config/api/filters.py:
```diff
@@ -0,0 +1,119 @@
+                # Attempt to convert the uuid string to a UUID object
+                config_template_uuid_obj = UUID(config_template_uuid)
+            except ValueError:
+                raise ValidationError({'config__templates': 'Invalid UUID format'})
```
remember to flag user facing strings as translatable, **always**.

Aryamanz29 replied:
noted.

nemesisdesign reviewed (changes requested):
Relation filters are missing, the rest looks good, once all is ready I will give it a deeper look.

Aryamanz29 commented:
> For all relations, can you add filters which show the available options in the UI? See the network topology example, eg:
> 
> For organization, can you try adding both organization as a relation and organization slug (as floorplan and location endpoints)? This should make it easier for the user to use whatever they have (whether it's the org ID or org slug).
> 
> We have added multi-tenant capabilities in these filters already if I am not mistaken, please read the and nearby sections extensively: [openwisp/openwisp-users#django-rest-framework-mixins](https://github.com/openwisp/openwisp-users#django-rest-framework-mixins).

@nemesisdesign I locally added the relation filters in the UI, but while testing, I encountered an issue that I believe is related to `openwisp-users`:

I think the `openwisp-users` have [implemented multi-tenant capabilities for DRF views, serializers, and Django admin filters](https://github.com/openwisp/openwisp-users#django-rest-framework-mixins), but they do not have it for [django-filters](https://django-filter.readthedocs.io/en/stable/). To double-check this, I tested an existing endpoint that displays relation options in the API filter, for example, **openwisp-network-topology** http://localhost:8000/api/v1/network-topology/node/. During testing, I found that the relationship fields are displayed in the filter UI without filtering by the organization that the user has access to, which breaks multi-tenancy.

#### I have created operator user that belongs to `org2` :

![Screenshot from 2023-03-20 16-42-01](https://user-images.githubusercontent.com/56113566/226323894-a2682387-e93b-41c4-8af4-c6f25eac542e.png)

#### In the filter options, it shows me the topologies of both organisations (default & org2) without filtering (which breaks multitency) :

![Screenshot from 2023-03-20 16-43-38](https://user-images.githubusercontent.com/56113566/226324063-72cb06eb-1be2-4dfd-9cea-639cb6608f83.png)

#### In the DRF create form and admin filter, it shows me the correct relation filter options (the same we want to achieve for django-filters) :

![Screenshot from 2023-03-20 16-43-54](https://user-images.githubusercontent.com/56113566/226324121-146ef0fc-76b0-4cd4-a9f3-b294eab8da9c.png)

![Screenshot from 2023-03-20 16-43-31](https://user-images.githubusercontent.com/56113566/226325627-ada8fe61-435f-4cf6-8e29-ef748998d883.png)

@nemesisdesign Now, the question is: what should we do? 

Should we complete this PR without relation options in the filter UI and then open a new issue in `openwisp-users` to add multitency capabilities to `django-filters` and update it here? Or should we first add multitency capabilities to `django-filters` and then update this PR?

nemesisdesign commented:
@Aryamanz29 let's complete this now.

Aryamanz29 commented:
#### `Update` : Added django-filters multi-tenancy (now both DRF and filter options looks similar)

![Screenshot from 2023-03-24 17-11-33](https://user-images.githubusercontent.com/56113566/227518427-b8ca4ee0-43da-484c-a99d-1d72a4210538.png)

![Screenshot from 2023-03-24 17-12-09](https://user-images.githubusercontent.com/56113566/227518373-fd8cebe0-1091-49f6-976e-914293be3634.png)

nemesisdesign commented on openwisp_controller/config/api/filters.py:
```diff
@@ -0,0 +1,145 @@
+class DeviceListFilter(BaseConfigAPIFilter):
+    # Using filter query param name `with_geo`
+    # which is similar to admin filter
+    with_geo = filters.BooleanFilter(
+        field_name='devicelocation', method='filter_devicelocation'
+    )
```
There is a problem here. The `config` app does not know about `geo`.
Look at how the admin filter is implemented, its code is in `openwisp_controller.geo`, not in `openwisp_controller.config`:

https://github.com/openwisp/openwisp-controller/blob/2b978723bf793aa40c88bd2d8f8e9bc8cd26c1bf/openwisp_controller/geo/admin.py#L85-L103

Adding this code here creates issue when `geo` is not installed.

Aryamanz29 replied:
@nemesisdesign I attempted to add DeviceLocationFilter with similar logic to what was used in `admin.py`, but it did not work. Therefore, I used a different approach in commit [`2abc007` (#748)](https://github.com/openwisp/openwisp-controller/pull/748/commits/2abc0071169317b21e5bd72ecf3d3b36f6d7b654), and it is now working as expected. Please see the details below:

#### When `geo` is not installed (device location filter is absent)

#### DRF browsable API

![Screenshot from 2023-04-24 15-40-17](https://user-images.githubusercontent.com/56113566/233969176-61c94f41-cd6c-428c-9bc4-8e4f26de01a9.png)

#### Admin
![Screenshot from 2023-04-24 15-40-29](https://user-images.githubusercontent.com/56113566/233969185-e3a1886a-314c-48ab-a64b-e153fdbfdb0e.png)

nemesisdesign commented on openwisp_controller/config/api/filters.py:
```diff
@@ -98,29 +94,32 @@ class DeviceListFilter(BaseConfigAPIFilter):
+            # Returns list of device that have devicelocation objects
+            return queryset.exclude(devicelocation__isnull=value)
+
+        with_geo = DeviceLocationFilter().filters['with_geo']
```
The app `controller.config` must not know anything about `controller.geo`.
I will push an alternative implementation that is acceptable and consistent within OpenWISP which I want you to look at and learn from.

nemesisdesign reviewed (changes requested):
I pushed the alternative solution after testing locally. Let's see if the build passes properly.

Since these filters are added by us and not provided by default and hence already tested in Django, we need a bit more tests to ensure they will not get messed with by future maintainers.
Please add at least one test for each filter (make sure the test fails if the code is broken on purpose).

Regarding `with_geo`, make sure the test for this filter is shipped in the `geo` app (ensure proper isolation).

Aryamanz29 commented:
> Since these filters are added by us and not provided by default and hence already tested in Django, we need a bit more tests to ensure they will not get messed with by future maintainers. Please add at least one test for each filter (make sure the test fails if the code is broken on purpose).

I have already added tests for all the implemented config app API filters in [config/tests/test_api.py](https://github.com/openwisp/openwisp-controller/pull/748/files#diff-56b0fb6f30ce93d1fd9d7195422a55bd97a97bdb35ab88758e352b1f54035c3f) 

> Regarding `with_geo`, make sure the test for this filter is shipped in the `geo` app (ensure proper isolation).

I have moved the `with_geo` device filter test to the `geo` app in [`cd008d5` (#748)](https://github.com/openwisp/openwisp-controller/pull/748/commits/cd008d5f3fb8c898e1f7f0420381c5d7316931bb)
